### PR TITLE
No-scheme relative path should be routed to HTTP handler only in browser

### PR DESCRIPTION
**Pull request by caisq**
* Currently, a call like tf.loadModel('./model.json') will be routed
  to tf.io.browserHTTPRequest regardless of whether the environment is
  browser or node.js. This will cause issues for node.js.
  This change makes it so that this routing of relative paths to
  browserHTTPRequest (using browser `fetch`) is done only in the
  browser. This implies that paths without a scheme is not supported.
  In other words, in node.js, a path given to tf.loadModel must have a
  scheme, e.g., tf.loadModel('file://./mode.json'),
  tf.loadModel('https://localhost/model.json').

Towards: https://github.com/tensorflow/tfjs/issues/343

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234)

### CODE REVIEW

**Comment by bileschi** [comment]
:lgtm_strong:

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-:-LEpIOGZ8xxgj8Jha6YG:b-jbsf8z)*

**Comment by dsmilkov** [comment]
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[src/models.ts, line 204 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-LEpvOT64_usIVpFAV27:-LEpvOT64_usIVpFAV28:bs80wvb) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/7df94c5d608cd01307e25e349001124493e66472/src/models.ts#L204)):*
> ```XML
>   if (typeof pathOrIOHandler === 'string') {
>     const handlers = io.getLoadHandlers(pathOrIOHandler);
>     if (handlers.length === 0 && ENV.get('IS_BROWSER')) {
> ```

does it make sense to have a "default" handler? Default handler gets called for relative paths. So if in the browser, register the httpHandler as the default handler. If node, register the fileHandler as the default handler. cc @nsthorat for thoughts.

---

*[src/models.ts, line 218 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-LEpvJEh6pAgmTIxDm_k:-LEpvJEiDaWggBtNVxhc:b1hyc8k) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/7df94c5d608cd01307e25e349001124493e66472/src/models.ts#L218)):*
> ```XML
>           `Cannot find any load handlers for path string '${pathOrIOHandler}'.`;
>       if (ENV.get('IS_NODE')) {
>         errorMsg +=
> ```

Can you add some unit tests that overwrite the IS_BROWSER ENV flag? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234)*

**Comment by nsthorat** [comment]
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[src/models.ts, line 204 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-LEpvOT64_usIVpFAV27:-LEpwWpQ3AU71lSCP_Mt:brhqij9) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/7df94c5d608cd01307e25e349001124493e66472/src/models.ts#L204)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

does it make sense to have a "default" handler? Default handler gets called for relative paths. So if in the browser, register the httpHandler as the default handler. If node, register the fileHandler as the default handler. cc @nsthorat for thoughts.
</blockquote></details>

I think that might make sense, but would be good to weigh against the complexity of having two handlers registered for the same thing (introducing priority, etc).

Just a musing: I don't really see a 3rd place where a relative URL would be registered, so I don't have a strong negative reaction against the proposed change here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234)*

**Comment by nsthorat** [comment]
Review status: :shipit: complete! 1 of 1 LGTMs obtained

---

*[src/models.ts, line 204 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-LEpvOT64_usIVpFAV27:-LEq0iKxBbqN-3hgTNC3:b-9mtqy1) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/7df94c5d608cd01307e25e349001124493e66472/src/models.ts#L204)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I think that might make sense, but would be good to weigh against the complexity of having two handlers registered for the same thing (introducing priority, etc).

Just a musing: I don't really see a 3rd place where a relative URL would be registered, so I don't have a strong negative reaction against the proposed change here.
</blockquote></details>

Sorry, I didn't see the description of the PR -- I don't think it's okay to force "file://" in node for relative paths.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234)*

**Comment by caisq** [comment]
Thank you all for the reviews so far. Let me know your thoughts, @nkreeger 

Two options for node.js + tf.loadModel

1) Let relative paths without URL schemes be treated as paths on the file system
2) Always require URL schemes in node.js + tf.loadModel. I.e., throw an error when there is no URL scheme.

This PR implements option 2.

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-:-LEqDVJr6KlXa6Ioo71q:b-2gqixn)*

**Comment by dsmilkov** [comment]
I'm definitely for #1 (Nikhil as well, based on his comment). Which is why I suggested an environment specific "default" handler that is responsible for handling relative paths.

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-:-LEtTl9J1wY9CspDpHJb:b-p0lerm)*

**Comment by nsthorat** [comment]
Yes -- I am also for #1.

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-:-LEtV8oo0C6bR8A7DOnZ:b-mi25zp)*

**Comment by bileschi** [comment]
This means that the same short program, executed in the two environments, will have different behavior.  This is not big problem, but we should be very clear about this re. documentation etc.

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/234#-:-LEtYsCy-qwxuKlk_lfF:brbnuov)*

**Comment by nsthorat** [comment]
Yes, this is true, but this is how relative paths work in general in JavaScript, so it's not a divergence from expectations.

**Comment by dsmilkov** [comment]
Hi Shanqing,

Just ran across this PR by chance. Let me know if you need more info/input from me to push this forward. Thanks!

**Comment by caisq** [comment]
@dsmilkov Sorry - I backburnered this a little due to other things at hand. It's also because users of tfjs-node have at least one route of saving/loading working for them already, namely file://. But I do plan to get this done soon.

**Comment by dsmilkov** [comment]
Makes sense - I trust your priority judgement :)

**Comment by jafaircl** [comment]
Any update on this? tfjs@0.14.1 seems to have broken file:// loading

**Comment by jafaircl** [comment]
@bileschi I looked on the js.tensorflow.org site [here](https://js.tensorflow.org/api/0.14.1/#Models-Loading)

**Comment by bileschi** [comment]
@tafsiri Does it make sense to add node-only API documentation here?  It seems to me the answer is yes, but it would be nice to have some sort of tag to indicate when API calls won't be available in some environments.

@nkreeger thoughts?

See also @jafaircl 's response on stackoverflow here : https://stackoverflow.com/questions/53031887/unable-to-load-python-trained-model-in-tensorflow-js

**Comment by tafsiri** [comment]
@bileschi From what I can gather this would just be documenting the behavior of the function (`tf.loadModel`) in node vs the browser? I think that makes sense here.

In cases where the function would not be available in a particular environment i believe the intent was to put it in an environment namespace like `tf.node` or `tf.browser` but i'm not sure if that applies here.

**Comment by caisq** [comment]
Since this is not super high priority. I'll close the PR and re-open it later in the monorepo.
